subsys/bluetooth/services/ots_oacp.c

```diff
@@ -0,0 +1,307 @@
+{
+	struct bt_gatt_ots *ots;
+
+	LOG_DBG("OACP Read procedure has been executed");
+
+	ots = CONTAINER_OF(l2cap_ctx, struct bt_gatt_ots, l2cap);
+	if (ots->cb.obj_read) {
+		ots->cb.obj_read(conn, ots->cur_obj->metadata.id);
+	}
+}
+
+static void oacp_read_proc_execute(struct bt_gatt_ots *ots)
+{
+	int err;
+	struct bt_gatt_ots_oacp_read_params *params =
+		&ots->cur_proc.read_params;
+
+	LOG_DBG("Executing Read procedure with offset: 0x%08X and "
+		"length: 0x%08X", params->offset, params->len);
+
+	if (!ots->cur_obj) {
+		LOG_ERR("Invalid Current Object on OACP Read procedure");
+		return;
+	}
+
+	ots->l2cap.tx_done = oacp_read_proc_executed;
+	err = bt_gatt_ots_l2cap_send(&ots->l2cap,
+				     &ots->cur_obj->content[params->offset],
+				     params->len);
+	if (err) {
+		LOG_ERR("L2CAP CoC error: %d while trying to execute OACP "
+			"Read procedure", err);
+	}
+}
+
+static void oacp_ind_cb(struct bt_conn *conn,
+			const struct bt_gatt_attr *attr,
+			u8_t err)
+{
+	struct bt_gatt_ots *ots = (struct bt_gatt_ots *) attr->user_data;
+
+	LOG_DBG("Received OACP Indication ACK with status: 0x%04X", err);
+
+	switch (ots->cur_proc.type) {
+		case BT_GATT_OTS_OACP_PROC_READ:
+			oacp_read_proc_execute(ots);
+			break;
+		case BT_GATT_OTS_OACP_PROC_CREATE:
+		case BT_GATT_OTS_OACP_PROC_DELETE:
+		case BT_GATT_OTS_OACP_PROC_CHECKSUM_CALC:
+		case BT_GATT_OTS_OACP_PROC_EXECUTE:
+		case BT_GATT_OTS_OACP_PROC_WRITE:
+		case BT_GATT_OTS_OACP_PROC_ABORT:
+		default:
+			break;
+	}
+}
+
+static int oacp_ind_send(const struct bt_gatt_attr *oacp_attr,
+			 enum bt_gatt_ots_oacp_proc_type req_op_code,
+			 enum bt_gatt_ots_oacp_res_code oacp_status)
+{
+	u8_t oacp_res[OACP_RES_MAX_SIZE];
+	u16_t oacp_res_len = 0;
+	struct bt_gatt_ots *ots = (struct bt_gatt_ots *) oacp_attr->user_data;
+
+	/* Encode OACP Response */
+	oacp_res[oacp_res_len++] = BT_GATT_OTS_OACP_PROC_RESP;
+	oacp_res[oacp_res_len++] = req_op_code;
+	oacp_res[oacp_res_len++] = oacp_status;
+
+	/* Prepare indication parameters */
+	memset(&ots->oacp_ind.params, 0, sizeof(ots->oacp_ind.params));
+	memcpy(&ots->oacp_ind.attr, oacp_attr, sizeof(ots->oacp_ind.attr));
+	ots->oacp_ind.params.attr = &ots->oacp_ind.attr;
+	ots->oacp_ind.params.func = oacp_ind_cb;
+	ots->oacp_ind.params.data = oacp_res;
+	ots->oacp_ind.params.len  = oacp_res_len;
+
+	LOG_DBG("Sending OACP indication");
+
+	return bt_gatt_indicate(NULL, &ots->oacp_ind.params);
+}
+
+ssize_t bt_gatt_ots_oacp_write(struct bt_conn *conn,
+			       const struct bt_gatt_attr *attr,
+			       const void *buf, u16_t len, u16_t offset,
+			       u8_t flags)
+{
+	enum bt_gatt_ots_oacp_res_code oacp_status;
+	struct bt_gatt_ots_oacp_proc oacp_proc;
+	struct bt_gatt_ots *ots = (struct bt_gatt_ots *) attr->user_data;
+
+	LOG_DBG("Object Action Control Point GATT Write Operation");
+
+	if (!ots->oacp_ind.is_enabled) {
+		LOG_WRN("OACP indications not enabled");
+		return BT_GATT_ERR(BT_ATT_ERR_CCC_IMPROPER_CONF);
+	}
+
+	if (offset != 0) {
```

Comment:
I don't see a check for 'procedure already in progress', is it missing?

Reply:
Yes, it is missing

Reply:
Is this still missing?

Reply:
No, it's not anymore. I added it at the very end of the ``oacp_read_proc_validate`` function if you want to have a look